### PR TITLE
io: add io_uring wrapper

### PR DESCRIPTION
**Pull request by rojkov**
Commit Message: io: add io_uring wrapper
Additional Description:
The wrapper can be used for asynchronous Block IO and Network IO in Linux
as an alternative to the standard POSIX system calls.

Risk Level: low, new code
Testing: unit tests added
Docs Changes: N/A
Release Notes: N/A, no changes visible to users
Platform Specific Features: io_uring wrapper for Linux

Contributes to #17922

### CODE REVIEW

**Comment by rojkov** [comment]
This is a smaller part of #19082.

/assign @antoniovicente

**Comment by jmarantz** [comment]
@wrowe @antoniovicente it looks like this is ready for another round of review?

File: test/common/io/io_uring_impl_test.cc
Comment (antoniovicente) [inline test/common/io/io_uring_impl_test.cc:78]: Use of epoll_wait to decide when to drive the IoUring seems like generic code that most uring users will needed.  Should this use of epoll_wait be moved to IoUringImpl ?

**Review by antoniovicente — COMMENTED** [review]
Interesting stuff.  I have some questions about the ultimate use of this new sophisticated mechanism.

File: test/common/io/io_uring_impl_test.cc
Comment (antoniovicente) [inline test/common/io/io_uring_impl_test.cc:1]: Please look at the test failures in CI:

[2021-12-28 13:52:10.547][15][critical][assert] [source/common/io/io_uring_impl.cc:34] assert failure: ret == 0. Details: unable to initialize io_uring: Operation not permitted

File: source/common/io/io_uring.h
Comment (rojkov) [reply source/common/io/io_uring.h:1]: This interface is tied tightly to syscalls which are specific to Linux and not generic enough. I made it abstract to ease unit testing of the io_uring-based iohandle (coming in followups).
But given [similar syscalls](https://windows-internals.com/i-o-rings-when-one-i-o-operation-is-not-enough/) are developed for Windows it might be useful to make it more generic. The problem is that I don't know how to design a proper abstraction at this stage.

File: source/common/io/io_uring.h
Comment (rojkov) [reply source/common/io/io_uring.h:21]: io_uring docs don't state any requirement regarding thread-safety though it probably should. In case of Envoy every IoUring is coupled with one libevent loop (it doesn't run its own) which is rarely posted from threads other than the one running it. So, I assumed that making IoUring thread local and calling its methods only from the thread it lives in simplifies everything. But perhaps it makes more sense to extend Dispatcher to own IoUrings.

File: source/common/io/io_uring.h
Comment (rojkov) [reply source/common/io/io_uring.h]: fixed

File: source/common/io/io_uring.h
Comment (rojkov) [reply source/common/io/io_uring.h:55]: Particularly this interface hides the io_uring syscalls and makes unit testing of the new iohandle (not in this PR) easier a bit. Originally I used a custom type containing the network iohandle instead of *void, but in case of Block IO a different type might be needed. Probably this *void in the interface could be avoided with a template.

I haven't thought of the possibility to implement an alternative Dispatcher, because its interface implies the file readiness model (i.e. Dispatcher::createFileEvent()), whereas io_uring is all about completeness of system operations. CompletionCb is an analog of FileReadyCb, but the former isn't called for a known file descriptor. A system operation and its context (including the arguments) are parts of a completeness event.

File: source/common/io/io_uring.h
Comment (rojkov) [reply source/common/io/io_uring.h]: added a clarification

File: source/exe/BUILD
Comment (rojkov) [reply source/exe/BUILD:108]: The problem is this wrapper doesn't get linked in Envoy's binary. It's not an extension and is not used by the new iohandle currently. If dropped it results in
```
Invalid build graph structure. deps(//source/...) != deps(//source/exe:envoy_main_common_with_core_extensions_lib) union deps(//source/extensions/...)
```
in `ci/run_envoy_docker.sh 'ci/do_ci.sh deps'`. The [original PR](https://github.com/envoyproxy/envoy/pull/19082) didn't need this dependency added.

File: test/common/io/io_uring_impl_test.cc
Comment (rojkov) [reply test/common/io/io_uring_impl_test.cc]: Thanks! This was a mistake. Fixed.

File: test/common/io/io_uring_impl_test.cc
Comment (rojkov) [reply test/common/io/io_uring_impl_test.cc:78]: Users don't need epoll_wait if they rely on libevent for event handling (via Dispatcher).

File: test/common/io/io_uring_impl_test.cc
Comment (rojkov) [reply test/common/io/io_uring_impl_test.cc:1]: This is complicated.

The test passes perfectly when run with Azure's managed workers: https://dev.azure.com/dmitryrojkov/meta-helloworld/_build/results?buildId=13&view=logs&j=8c169225-0ae8-53bd-947f-07cb81846cb5&t=6b0ace90-28b2-51d9-2951-b1fd35e67dec
It fails if RBE is active. I suspect that io_uring syscalls are not allow-listed on the workers used as RBEs. Do you happen to know how to reach out to people running those builders?

File: source/common/io/io_uring.h
Comment (rojkov) [reply source/common/io/io_uring.h:41]: added

File: test/common/io/io_uring_impl_test.cc
Comment (antoniovicente) [reply test/common/io/io_uring_impl_test.cc:1]: cc @yanavlasov who may have some thoughts.

File: test/common/io/io_uring_impl_test.cc
Comment (antoniovicente) [reply test/common/io/io_uring_impl_test.cc:78]: If the intended use is together with the dispatcher, perhaps this test should use the dispatcher for event handling.

File: source/common/io/io_uring.h
Comment (antoniovicente) [reply source/common/io/io_uring.h:55]: I think it's ok to use void* for now, we may revisit this later as our understanding of this module improves.

I'ld like to better understand how io_uring and dispatcher should interact.  I could see some threads being exclusively one or the other.  I'm not sure about how they would coexist in a thread while avoiding starvation.

File: source/common/io/io_uring.h
Comment (antoniovicente) [reply source/common/io/io_uring.h:21]: It may make sense for the dispatcher to own the io_urings eventually instead of providing a getOrCreate factory method.  We can get to that as development continues.

**Comment by antoniovicente** [comment]
Blocked on tests on RBE.  Could use some help from @yanavlasov 

/wait

File: source/common/io/io_uring.h
Comment (rojkov) [reply source/common/io/io_uring.h:55]: I see no problem of multiple io_urings coexisting in one dispatcher: all updates in their completion queues are reflected in their respective event notification file descriptors (`man eventfd`). So, different callback instances are called for different eventfds, only one registered callback per io_uring. The dispatcher is supposed to call these callbacks in the same manner as in the case of connection callbacks.

I have rewritten the tests to run the callbacks through Dispatcher::run().

Though I fail to come up with a simple enough API for Dispatcher to accommodate multiple io_urings if go this way. If we assume only one io_uring per Dispatcher then extending Dispatcher with something like `void Dispatcher::registerCompletionEvent(CompletionCb cb)` and `IoUring& Dispatcher::getIoUring()` would be enough, `IoUring::registerEventfd()` could be dropped, since all manipulations with eventfd would be incapsulated in Dispatcher.

For the case of multiple io_urings `Dispatcher::getIoUring()` and `Dispatcher::registerCompletionEvent()` should distinguish io_urings somehow. Maybe something like the following could work:

```c++
using IoUringId = uint32_t;

class DispatcherImpl {
public:
  ...
  IoUringId addIoUring(const IoUringFactory& factory);
  void registerCompletionEvent(IoUringId id, CompletionCb cb);
  IoUring& getIoUring(IoUringId id);
  ...

private:
  std::vector<IoUring> io_urings_;
};
```

File: test/common/io/io_uring_impl_test.cc
Comment (rojkov) [reply test/common/io/io_uring_impl_test.cc]: Good point. The test doesn't include `sys/epoll.h` now, event handlers are run by dispatcher instances.

**Review by yanavlasov — CHANGES_REQUESTED** [review]
/wait-any

File: source/common/io/io_uring.h
Comment (rojkov) [reply source/common/io/io_uring.h:1]: IIUC #19082 does exactly that - an IoHandle as a boostrap extension. This PR is an attempt to enable io_uring also for file IO. Additionally Matt [proposed](https://github.com/envoyproxy/envoy/pull/19082#issuecomment-981977371) to switch to io_uring whenever it's available without enabling it explicitly as an extension.

File: source/common/io/io_uring_impl.cc
Comment (rojkov) [reply source/common/io/io_uring_impl.cc:12]: This is interesting. I'll try that.

**Comment by rojkov** [comment]
/wait-any

Hopefully, will add commits with the mentioned bootstrap extension next week.

File: source/common/io/io_uring_impl.cc
Comment (rojkov) [reply source/common/io/io_uring_impl.cc]: I've made the factory to use Envoy's thread local storage. It works for this PR. But then I ported the change to the [draft PR](https://github.com/envoyproxy/envoy/pull/19082) with io_uring-backed iohandle and its integration test started to fail because the client connection created in the test_thread wants to use IoUring too. test_thread is not registered with TLS -> its slot is not set -> crash. Can test_thread be registered properly?

Also what's the benefit of using Envoy's TLS here?

**Comment by rojkov** [comment]
@yanavlasov gentle ping. This is ready for another pass.

**Comment by wrowe** [comment]
@htuch Would you also take a look on behalf of deps, w.r.t. https://github.com/axboe/liburing - this otherwise would lgtm

File: bazel/repository_locations.bzl
Comment (wrowe) [inline bazel/repository_locations.bzl:126]: Note https://github.com/envoyproxy/envoy/pull/19082#pullrequestreview-813623549 which seems to apply here as well.

**Comment by htuch** [comment]
I'm not seeing a major CVE history, but also this project lacks security policy. I think the GitHub flow is not canonical and this follows a LKML approach to code review, see https://lore.kernel.org/io-uring/. There do seem to be numerous commits from the author that don't have a review trail. OSSF scorecard indicates lack of branch protection, CI and fuzzing. The author is also the author of the kernel side of this mechanism.

TBH, I would not be a fan of adding this if it wasn't for the substantial benefit it brings in terms of convenience, which is the reason I gather we're using it. I gather this is the standard io_uring library in C-languages.

I think in this case the benefits outweigh the downsides and we're seeing the usual impedance mismatch between Linux kernel style development flows and GitHub.

@moderation for a second opinion.

**Comment by moderation** [comment]
The OSSF Scorecard result has worsened with recent updates - `scorecard --repo=https://github.com/axboe/liburing`
```
RESULTS
-------
Aggregate score: 4.6 / 10

Check scores:
|---------|------------------------|--------------------------------|-----------------------------------------------------------------------------------|
|  SCORE  |          NAME          |             REASON             |                             DOCUMENTATION/REMEDIATION                             |
|---------|------------------------|--------------------------------|-----------------------------------------------------------------------------------|
| 10 / 10 | Binary-Artifacts       | no binaries found in the repo  | https://github.com/ossf/scorecard/blob/main/docs/checks.md#binary-artifacts       |
|---------|------------------------|--------------------------------|-----------------------------------------------------------------------------------|
| 0 / 10  | Branch-Protection      | branch protection not enabled  | https://github.com/ossf/scorecard/blob/main/docs/checks.md#branch-protection      |
|         |                        | on development/release         |                                                                                   |
|         |                        | branches                       |                                                                                   |
|---------|------------------------|--------------------------------|-----------------------------------------------------------------------------------|
| 0 / 10  | CI-Tests               | 0 out of 4 merged PRs          | https://github.com/ossf/scorecard/blob/main/docs/checks.md#ci-tests               |
|         |                        | checked by a CI test -- score  |                                                                                   |
|         |                        | normalized to 0                |                                                                                   |
|---------|------------------------|--------------------------------|-----------------------------------------------------------------------------------|
| 0 / 10  | CII-Best-Practices     | no badge detected              | https://github.com/ossf/scorecard/blob/main/docs/checks.md#cii-best-practices     |
|---------|------------------------|--------------------------------|-----------------------------------------------------------------------------------|
| 1 / 10  | Code-Review            | GitHub code reviews found for  | https://github.com/ossf/scorecard/blob/main/docs/checks.md#code-review            |
|         |                        | 4 commits out of the last 30   |                                                                                   |
|         |                        | -- score normalized to 1       |                                                                                   |
|---------|------------------------|--------------------------------|-----------------------------------------------------------------------------------|
| 10 / 10 | Contributors           | 20 different companies found   | https://github.com/ossf/scorecard/blob/main/docs/checks.md#contributors           |
|         |                        | -- score normalized to 10      |                                                                                   |
|---------|------------------------|--------------------------------|-----------------------------------------------------------------------------------|
| 10 / 10 | Dangerous-Workflow     | no dangerous workflow patterns | https://github.com/ossf/scorecard/blob/main/docs/checks.md#dangerous-workflow     |
|         |                        | detected                       |                                                                                   |
|---------|------------------------|--------------------------------|-----------------------------------------------------------------------------------|
| 0 / 10  | Dependency-Update-Tool | no update tool detected        | https://github.com/ossf/scorecard/blob/main/docs/checks.md#dependency-update-tool |
|---------|------------------------|--------------------------------|-----------------------------------------------------------------------------------|
| 0 / 10  | Fuzzing                | project is not fuzzed          | https://github.com/ossf/scorecard/blob/main/docs/checks.md#fuzzing                |
|---------|------------------------|--------------------------------|-----------------------------------------------------------------------------------|
| 10 / 10 | License                | license file detected          | https://github.com/ossf/scorecard/blob/main/docs/checks.md#license                |
|---------|------------------------|--------------------------------|-----------------------------------------------------------------------------------|
| 10 / 10 | Maintained             | 30 commit(s) out of 30 and 14  | https://github.com/ossf/scorecard/blob/main/docs/checks.md#maintained             |
|         |                        | issue activity out of 30 found |                                                                                   |
|         |                        | in the last 90 days -- score   |                                                                                   |
|         |                        | normalized to 10               |                                                                                   |
|---------|------------------------|--------------------------------|-----------------------------------------------------------------------------------|
| ?       | Packaging              | no published package detected  | https://github.com/ossf/scorecard/blob/main/docs/checks.md#packaging              |
|---------|------------------------|--------------------------------|-----------------------------------------------------------------------------------|
| 9 / 10  | Pinned-Dependencies    | dependency not pinned by hash  | https://github.com/ossf/scorecard/blob/main/docs/checks.md#pinned-dependencies    |
|         |                        | detected -- score normalized   |                                                                                   |
|         |                        | to 9                           |                                                                                   |
|---------|------------------------|--------------------------------|-----------------------------------------------------------------------------------|
| 0 / 10  | SAST                   | SAST tool is not run on all    | https://github.com/ossf/scorecard/blob/main/docs/checks.md#sast                   |
|         |                        | commits -- score normalized to |                                                                                   |
|         |                        | 0                              |                                                                                   |
|---------|------------------------|--------------------------------|-----------------------------------------------------------------------------------|
| 0 / 10  | Security-Policy        | security policy file not       | https://github.com/ossf/scorecard/blob/main/docs/checks.md#security-policy        |
|         |                        | detected                       |                                                                                   |
|---------|------------------------|--------------------------------|-----------------------------------------------------------------------------------|
| ?       | Signed-Releases        | no releases found              | https://github.com/ossf/scorecard/blob/main/docs/checks.md#signed-releases        |
|---------|------------------------|--------------------------------|-----------------------------------------------------------------------------------|
| 0 / 10  | Token-Permissions      | non read-only tokens detected  | https://github.com/ossf/scorecard/blob/main/docs/checks.md#token-permissions      |
|         |                        | in GitHub workflows            |                                                                                   |
|---------|------------------------|--------------------------------|-----------------------------------------------------------------------------------|
| 10 / 10 | Vulnerabilities        | no vulnerabilities detected    | https://github.com/ossf/scorecard/blob/main/docs/checks.md#vulnerabilities        |
|---------|------------------------|--------------------------------|-----------------------------------------------------------------------------------|
```
Detailed report at [liburing_ossf_scorecard.txt](https://github.com/envoyproxy/envoy/files/8112758/liburing_ossf_scorecard.txt)

Is this really low risk considering it will be used when supported on Linux by default? Are code reviews handled in the mailing list at https://lore.kernel.org/io-uring/?

**Comment by htuch** [comment]
The main thing is I'm not sure if the primary author of this library is doing code reviews for their own code, e.g. take a commit like https://github.com/axboe/liburing/commit/d84c29b19ed0b130000619cff40141bb1fc3615b, is there a corresponding public code review?

For a core library, I think this is pretty important, we've been making a concerted effort to move away from nghttp2 for this reason; OSS projects that don't have a strong process around public review for all commits, regardless of author, should be treated carefully from a supply chain angle.

**Comment by htuch** [comment]
@rojkov if you can do that without undue pain that could be good. With all these supply chain considerations, it's a tradeoff between the convenience and risk; if it ends up being more error prone to not use a library there is a risk there as well. So, with the above concerns noted, happy to try and find the right balance.

**Comment by rojkov** [comment]
I tried to use syscalls directly instead of liburing and very soon realized that what I come up with is a copy-paste of the lib basically. And rather complex one which we definitely don't want to maintain. So, I'd propose to proceed with liburing.

**Comment by htuch** [comment]
Sounds like a good faith effort to explore alternatives and this is the least bad option, at least if we want to target performance, so no objections.

**Comment by wrowe** [comment]
Last call, it sounds like concerns of @htuch , @moderation, @yanavlasov and @antoniovicente are all addressed? If I don't hear otherwise, I'll stamp the deps and merge this Friday morning.

File: api/envoy/extensions/io/io_uring/v3/io_uring.proto
Comment (htuch) [inline api/envoy/extensions/io/io_uring/v3/io_uring.proto:1]: Why isn't this in the bootstrap extension hierarchy? Looking below, I think this is what it is. This hierarchy is introduced in https://github.com/envoyproxy/envoy/pull/19467/files, it seems out bootstrap extensions have been adhoc scattered so far :(

File: source/common/io/io_uring_impl.cc
Comment (htuch) [inline source/common/io/io_uring_impl.cc:22]: This is a bit strange, this is both in core but also an extension?

Comment (rojkov) [reply source/common/io/io_uring_impl.cc:22]: Uh, right. Originally it was not an extension. It was turned into an extension in a later commit to avoid `static thread_local` and to use the TLS API instead as suggested.

Now I'm confused. If I keep the code to be an extension io_uring won't be usable in core. Otherwise the TLS API seems to be unavailable. I'd rather drop TLS and bootstrap extension as it was originally.

Comment (htuch) [reply source/common/io/io_uring_impl.cc:22]: Now _I'm_ confused, so you're not the only one :) Why can't we use TLS in extensions? Why does this need to be used directly by core?

Comment (rojkov) [reply source/common/io/io_uring_impl.cc:22]: > Why does this need to be used directly by core?

[This comment](https://github.com/envoyproxy/envoy/pull/19082#issuecomment-981977371) is the main reason.

> Why can't we use TLS in extensions?

That was about (un)availability of TLS in core, not extensions. This concern is moot probably. Still I'm not sure why using TLS is more beneficial than `static thread_local` in this case, because I bumped into [other problems with integration testing](https://github.com/envoyproxy/envoy/pull/19339#discussion_r802882948).

Comment (htuch) [reply source/common/io/io_uring_impl.cc:22]: OK, then let's make it core. We can use TLS from non-extensions, it's used all over core, e.g. see https://github.com/envoyproxy/envoy/blob/1f8da8e0fffed1c5b387ca2a39e6f5380bc1613f/source/common/upstream/cluster_manager_impl.h#L726.

I think one of the nice things about using Envoy TLS is you get TLS lifecycle management, which should in theory actually work better for testing etc.

Comment (rojkov) [reply source/common/io/io_uring_impl.cc]: Thanks!

So far my problem with Envoy TLS in testing has been that the thread controlling integration tests doesn't initialize its TLS slot hence lacks io_uring which is needed to open a client connection. I'll experiment with enabling TLS in the testing thread.

The bootstrap extension is dropped now.

**Comment by adisuissa** [comment]
@htuch @yanavlasov PTAL

File: api/versioning/BUILD
Comment (htuch) [inline api/versioning/BUILD:158]: Remove

**Review by htuch — COMMENTED** [review]
Other than nit on extension, no longer any objection on API / extension side.

File: api/versioning/BUILD
Comment (rojkov) [reply api/versioning/BUILD]: Oops, removed.

**Comment by alyssawilk** [comment]
@wrowe this has been waiting on your dependency stamp for 8 days?

File: source/common/io/io_uring_impl.cc
Comment (rojkov) [reply source/common/io/io_uring_impl.cc]: I turned the extension back to core per Harvey's comment. Yet Envoy's TLS is used now instead of `thread_local`.

**Comment by rojkov** [comment]
Yep, I've just marked the addressed feedback from Yan as resolved.

**Review by yanavlasov — APPROVED** [review]

**Pull request merged**
